typescript/src/v0/Mcap0Writer.ts

```diff
@@ -0,0 +1,415 @@
+import { IWritable } from "../common/IWritable";
+import { ChunkBuilder } from "./ChunkBuilder";
+import { Mcap0RecordBuilder } from "./Mcap0RecordBuilder";
+import { Opcode } from "./constants";
+import {
+  Schema,
+  Channel,
+  Message,
+  Header,
+  Attachment,
+  Chunk,
+  ChunkIndex,
+  AttachmentIndex,
+  MetadataIndex,
+  SummaryOffset,
+  Metadata,
+  Statistics,
+} from "./types";
+
+type Mcap0WriterOptions = {
+  writable: IWritable;
+  useStatistics?: boolean;
+  useSummaryOffsets?: boolean;
+  useChunks?: boolean;
+  repeatSchemas?: boolean;
+  repeatChannels?: boolean;
+  useAttachmentIndex?: boolean;
+  useMetadataIndex?: boolean;
+  useMessageIndex?: boolean;
+  useChunkIndex?: boolean;
+};
+
+/**
+ * Mcap0Writer provides an interface for writing messages to MCAP files.
+ *
+ * NOTE: callers must wait on any method call to complete before calling another
+ * method. Calling a method before another has completed will result in a corrupt
+ * MCAP file.
+ */
+export class Mcap0Writer {
+  private static MIN_CHANNEL_ID = 1; //FIXME- only for conformance tests
+  private static MIN_SCHEMA_ID = 1;
+  private writable: IWritable;
+  private recordWriter = new Mcap0RecordBuilder();
+  private schemas = new Map<number, Schema>();
+  private channels = new Map<number, Channel>();
+  private writtenSchemaIds = new Set<number>();
+  private writtenChannelIds = new Set<number>();
+  private chunkBuilder: ChunkBuilder | undefined;
+
+  private statistics: Statistics | undefined;
+  private useSummaryOffsets: boolean;
+  private repeatSchemas: boolean;
+  private repeatChannels: boolean;
+
+  // indices
+  private chunkIndices: ChunkIndex[] | undefined;
+  private attachmentIndices: AttachmentIndex[] | undefined;
+  private metadataIndices: MetadataIndex[] | undefined;
+
+  constructor({
```

Comment:
I am not a fan of de-structuring inline with function declaration. I find it hard to read. If it doesn't make much difference to you I'd prefer if this was split into the function body.